wrappers/winrt/BarcodeReader.cpp

```diff
@@ -1,124 +1,234 @@
-		std::shared_ptr<GenericLuminanceSource> luminance;
-		switch (bitmap->BitmapPixelFormat)
-		{
-		case BitmapPixelFormat::Gray8:
-			luminance = std::make_shared<GenericLuminanceSource>(cropLeft, cropTop, cropWidth, cropHeight, inBytes, inBuffer->GetPlaneDescription(0).Stride);
-			break;
-		case BitmapPixelFormat::Bgra8:
-			luminance = std::make_shared<GenericLuminanceSource>(cropLeft, cropTop, cropWidth, cropHeight, inBytes, inBuffer->GetPlaneDescription(0).Stride, 4, 2, 1, 0);
-			break;
-		case BitmapPixelFormat::Rgba8:
-			luminance = std::make_shared<GenericLuminanceSource>(cropLeft, cropTop, cropWidth, cropHeight, inBytes, inBuffer->GetPlaneDescription(0).Stride, 4, 0, 1, 2);
-			break;
-		default:
-			throw std::runtime_error("Unsupported format");
-		}
-		return std::make_shared<HybridBinarizer>(luminance);
-	}
-	else
-	{
-		throw std::runtime_error("Failed to read bitmap's data");
-	}
-}
-
-static Platform::String^ ToPlatformString(const std::wstring& str)
-{
-	return ref new Platform::String(str.c_str(), (unsigned)str.length());
-}
-
-static Platform::String^ ToPlatformString(const std::string& str)
-{
-	auto ptr = (const uint8_t*)str.data();
-	return ToPlatformString(std::wstring(ptr, ptr + str.length()));
-}
-
-ReadResult^
-BarcodeReader::Read(SoftwareBitmap^ bitmap, int cropWidth, int cropHeight)
-{
-	try {
-		auto binImg = CreateBinaryBitmap(bitmap, cropWidth, cropHeight);
-		auto result = m_reader->read(*binImg);
-		if (result.isValid()) {
-			return ref new ReadResult(ToPlatformString(ZXing::ToString(result.format())), ToPlatformString(result.text()));
-		}
-	}
-	catch (const std::exception& e) {
-		OutputDebugStringA(e.what());
-	}
-	catch (...) {
-		
-	}
-	return nullptr;
-}
-
-} // ZXing
+#endif
+
+#include "BarcodeReader.h"
+
+#include "BarcodeFormat.h"
+#include "DecodeHints.h"
+#include "GenericLuminanceSource.h"
+#include "HybridBinarizer.h"
+#include "MultiFormatReader.h"
+#include "ReadResult.h"
+#include "Result.h"
+
+#include <algorithm>
+#include <MemoryBuffer.h>
+#include <stdexcept>
+#include <wrl.h>
+
+using namespace Microsoft::WRL;
+using namespace Windows::Foundation;
+using namespace Windows::Graphics::Imaging;
+
+namespace ZXing {
+
+BarcodeReader::BarcodeReader(bool tryHarder, bool tryRotate, const Platform::Array<BarcodeType>^ types)
+{
+	DecodeHints hints;
+	hints.setShouldTryHarder(tryHarder);
+	hints.setShouldTryRotate(tryRotate);
+
+	if (types != nullptr && types->Length > 0) {
+		std::vector<BarcodeFormat> barcodeFormats;
+		for (BarcodeType type : types) {
+			barcodeFormats.emplace_back(BarcodeReader::ConvertRuntimeToNative(type));
+		}
+		hints.setPossibleFormats(barcodeFormats);
+	}
+
+	m_reader.reset(new MultiFormatReader(hints));
+}
+
+BarcodeReader::BarcodeReader(bool tryHarder, bool tryRotate)
+{
+	DecodeHints hints;
+	hints.setShouldTryHarder(tryHarder);
+	hints.setShouldTryRotate(tryRotate);
+	m_reader.reset(new MultiFormatReader(hints));
+}
+
+BarcodeReader::BarcodeReader(bool tryHarder)

```

Comment:
I'm seeing an issue during compilation when It try to use a delegating constructor. I believe it's related to an implicit dispose() being called somewhere in the templates. I'm still digging into what can be done about this but if you have some insight, it would be appreciated.

Header/Classfile combination

```
Header:

BarcodeReader(bool tryHarder, bool tryRotate); 

Class file:

BarcodeReader::BarcodeReader(bool tryHarder, bool tryRotate) : BarcodeReader(tryHarder, tryRotate, nullptr) {}
```
Error Output:
```
1>C:\projects\zxing-cpp\wrappers\winrt\BarcodeReader.cpp(60): error C2039: '<Dispose>': is not a member of 'ZXing::BarcodeReader'
1>c:\projects\zxing-cpp\wrappers\winrt\BarcodeReader.h(44): note: see declaration of 'ZXing::BarcodeReader'
1>C:\projects\zxing-cpp\wrappers\winrt\BarcodeReader.cpp(60): error C2248: 'Microsoft::WRL::RuntimeClass<Microsoft::WRL::RuntimeClassFlags<2>,IUnknown>::operator =': cannot access private member declared in class 'Microsoft::WRL::RuntimeClass<Microsoft::WRL::RuntimeClassFlags<2>,IUnknown>'
1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\winrt\wrl\implements.h(2170): note: see declaration of 'Microsoft::WRL::RuntimeClass<Microsoft::WRL::RuntimeClassFlags<2>,IUnknown>::operator ='
1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\winrt\wrl\event.h(441): note: see declaration of 'Microsoft::WRL::RuntimeClass<Microsoft::WRL::RuntimeClassFlags<2>,IUnknown>'

```



When declaring the implementation directly in the header:
```
BarcodeReader(bool tryHarder) : BarcodeReader(tryHarder, tryHarder, nullptr) {};
```

Error Output:
```

1>c:\projects\zxing-cpp\wrappers\winrt\BarcodeReader.h(49): error C2039: '<Dispose>': is not a member of 'ZXing::BarcodeReader'
1>c:\projects\zxing-cpp\wrappers\winrt\BarcodeReader.h(44): note: see declaration of 'ZXing::BarcodeReader'
```

Reply:
Had to change visibility on the destructor and declare it virtual

Reply:
You are right! I'm not sure what is going on neither, from the message, its seems that the compiler generates code using `operator=` which is of course not available for ref class. I've pushed a commit to move initialization to a private member method. I also fix a warning with wchar_t to char conversion.

Reply:
I remember that make destructor public virtual implies that BarcodeReader will become IDisposable on C# side. That is an API change that will not be transparent.

Reply:
I did manage to get it working with Delegating constructors by making the destructor virtual and public. Do you prefer we go that route and use the constructors or do you want to stick with the init ?

Reply:
@William-Froelich, See my answer just just above. If it's ok for you, I would go with init().

Reply:
I'm fine with the proposed changes.